FIX: Updating digitizer point decimation to units of meters

kdoelling1919 opened the pull request:
Updating automatic decimation of digitizer headshape to units of m rather than mm in keeping with recent updates. Fixes #3746 

@teonbrooks @christianbrodbeck @Eric89GXL

christianbrodbeck commented:
Looks good to me

Eric89GXL commented:
LGTM +1 for merge once CIs are happy. It would be nice to have a test that would have caught this but I assume it's difficult due to being a GUI function.

agramfort commented:
does it mean it will break with old files?

Eric89GXL commented:
I think it means it's currently broken for all files, i.e. we changed units in not all places we needed to somewhere...

agramfort commented:
arfff ok :(

christianbrodbeck commented:
Yes, @Eric89GXL is right, it's currently broken for all files, I missed this conditional when changing units. 

It's not a GUI function so a test could be added, all that would have to be done is generate >10000 HSP points.

Eric89GXL commented:
@kdoelling1919 can you do it? `mne.surface._get_ico_surface(5)['rr']` will get you 10242 points arranged on a unit sphere, all you should need to do is scale them and add some simple test.

kdoelling1919 commented:
I think I should be able to handle that. I haven't written any tests before but it sounds reasonable enough. 

I'm not certain though what the test should be over? Should I run the test over read_raw_kit replacing an hsp file with the sphere and checking that it is still a similar sphere after decimation? 

Is there some minimum number of points for which we can expect a head shape to be useful? It seems like that would also be a very useful check in general.

Eric89GXL commented:
> I'm not certain though what the test should be over?

Ideally yeah you would use a public function, but in this case I think testing the private one is probably okay.

> Is there some minimum number of points for which we can expect a head shape to be useful?

It depends on the eventual application / use case -- for example for sensor space analyses they don't matter at all. So I'd say the check should be delayed until something that uses the points (head<->MRI alignment, sphere fitting, etc.) occurs.

christianbrodbeck commented:
>> I'm not certain though what the test should be over?

> Ideally yeah you would use a public function, but in this case I think testing the private one is probably okay.

To actually test the whole workflow wouldn't you generate a temporary HSP file and use `mne.io.read_raw_kit`?

Eric89GXL commented:
Yeah that would be a more complete version (and what he suggested), I'm fine with that if you think it's worth the extra effort

christianbrodbeck commented:
I think it would be safer, the change in the units happened early in `read_raw_kit` so a test that would have skipped the step reading from the text file might have missed it...

The test could probably be combined with the other read_raw_kit tests with little added code, just supply a HSP file with >10000 points

kdoelling1919 commented:
OK, I added a preliminary test using the `_get_ico_surface(5)['rr']` sphere and ran it through `mne.io.read_raw_kit()` and added some very simple tests. Please let me know if there are any other checks I should add.

christianbrodbeck commented on mne/io/kit/tests/test_kit.py:
```diff
@@ -185,4 +186,24 @@ def test_hsp_elp():
+    hsp_rad = np.mean(dist)
+    hsp_dec_rad = np.mean(dist_dec)
+
+    assert_array_almost_equal(hsp_rad, hsp_dec_rad, decimal=3)
```
Since these are both scalars wouldn't it be simpler to use `nose`'s `assert_almost_equal`?

kdoelling1919 replied:
Probably! I didn't know about `nose`

christianbrodbeck replied:
Oh, it's in `nose.tools`

teonbrooks replied:
you can also use `numpy.testing.assert_array_almost_equal`. i think they are basically the same

christianbrodbeck replied:
numpy functions have more overhead for coercing the arguments to arrays -- not that it matters much for a single call

christianbrodbeck replied:
although in this case it might actually matter pretty soon
```
In [3]: from numpy.testing import assert_array_almost_equal

In [4]: from nose.tools import assert_almost_equal

In [5]: timeit assert_array_almost_equal(1, 1)
10000 loops, best of 3: 75.1 µs per loop

In [6]: timeit assert_almost_equal(1, 1)
The slowest run took 66.25 times longer than the fastest. This could mean that an intermediate result is being cached.
10000000 loops, best of 3: 151 ns per loop
```

Eric89GXL replied:
We typically use the `numpy` versions but I don't think it will matter too much for our use cases

christianbrodbeck commented on mne/io/kit/tests/test_kit.py:
```diff
@@ -185,4 +186,24 @@ def test_hsp_elp():
+    hsp /= 10
+
+    # read in raw data using spherical hsp, and extract new hsp
+    raw = read_raw_kit(sqd_path, mrk_path, elp_txt_path, hsp)
```
I wonder if it would be worth writing the points to a temporary `.txt` file too, since the actual unit conversion happens in `_read_dig_points()` which is not called when the points are specified as array.

teonbrooks commented on mne/io/kit/kit.py:
```diff
@@ -504,7 +504,7 @@ def _set_dig_kit(mrk, elp, hsp):
     n_pts = len(hsp)
     if n_pts > KIT.DIG_POINTS:
-        hsp = _decimate_points(hsp, res=5)
+        hsp = _decimate_points(hsp, res=0.005)
```
related, @christianbrodbeck, should we update the default value for _decimate_points and say that it should be based on `m` or we could say in general, the appropriate scale of your data

christianbrodbeck replied:
I think it makes sense for it to be unit-agnostic as it currently is, since you enter both the values and the grid spacing.

teonbrooks commented on mne/io/kit/tests/test_kit.py:
```diff
@@ -185,4 +186,24 @@ def test_hsp_elp():
+    """Test decimation of digitizer headshapes with too many points. """
+    # create fake head points from sphere, and scale to similar numbers as hsp
+    hsp = _get_ico_surface(5)['rr']
+    hsp /= 10
```
are you rescaling the values here?

teonbrooks replied:
@Eric89GXL do you know the units of the surface? is it already in `m`?

kdoelling1919 replied:
The sphere is basically just a unit sphere from what I could tell.

teonbrooks replied:
oh ok, then I don't think the rescaling is needed then

teonbrooks commented on mne/io/kit/tests/test_kit.py:
```diff
@@ -185,4 +186,24 @@ def test_hsp_elp():
+    raw = read_raw_kit(sqd_path, mrk_path, elp_txt_path, hsp)
+    hsp_dec = np.array([dig['r'] for dig in raw.info['dig']])
+
+    # with 10242 points and decimation at 5 mm, hsp_dec should be over 5000
```
why would we expect there to be over 5000 points after decimation?

kdoelling1919 replied:
It just so happens that decimating with current parameters (5 mm boxes) of the sphere with 10242 points goes down to 5000+ points. I wanted to have a check that would have caught the error that's being fixed here. Agreed it's not the best choice (I.e. it could fail if at some point we decide to change the resolution parameter) but it was the best I could think of.

Eric89GXL replied:
sounds good, but please modify the comment so the next person modifying the code has more guidance

agramfort commented:
no objection to merge

kdoelling1919 commented:
Hi all,

I would very much like to make these changes but my computer won't boot at all despite much troubleshooting and I'm in Thailand for the week so I won't be able to do much fixing before I get back. If someone wants to incorporate some of the changes themselves they are welcome to! Otherwise we'll have to wait till next week.

Sorry!
Keith

Eric89GXL commented:
No rush from my end @kdoelling1919 I think next week would be fine. I hope your computer comes to its senses

kdoelling1919 commented:
OK I think i've covered all of the changes requested here. (For those interested, my computer issues had to do with a faulty charger which claimed to be charging my computer but for the most part wasn't)

Let me know if there's anything else I missed. Otherwise, feel free to merge!

agramfort commented on mne/io/kit/tests/test_kit.py:
```diff
@@ -185,4 +188,24 @@ def test_hsp_elp():
+    # with 10242 points and _decimate_points set to resolution of 5 mm, hsp_dec
+    # should be a bit over 5000 points. If not, something is wrong or
+    # decimation resolution has been purposefully changed
+    assert len(hsp_dec) > 5000
```
use assert_true not assert

agramfort commented:
@teonbrooks @christianbrodbeck please co-merge if you're happy

christianbrodbeck commented on mne/io/kit/tests/test_kit.py:
```diff
@@ -185,4 +188,24 @@ def test_hsp_elp():
+    """Test decimation of digitizer headshapes with too many points. """
+    # load headshape and convert to meters
+    hsp = np.loadtxt(sphere_hsp_txt_path)
+    hsp /= 1000
```
It might be better to generate the sphere head shape file during the test (as you did before), and save it in a temporary director before loading again? Including it as a text file in the repository adds 765 KB of data. I am not sure at what size data files for tests become a problem but seem to remember that the issue was raised at some point (@Eric89GXL ?).

Eric89GXL replied:
Do not add 700+kB for a sphere of points. We can generate them easily. Is it slow or something? If so just do a `meshgrid` with azimuth/polar and `sph_to_cart` and it will be fast.

christianbrodbeck replied:
You can use `_TempDir()` like here https://github.com/mne-tools/mne-python/blob/474b8d2a6e9df06099aff1012a2c0722549ee72d/mne/tests/test_coreg.py#L25 to quickly generate a temporary directory for the test

kdoelling1919 replied:
I used `NamedTemporaryFile` from the `tempfile` module before seeing this comment. Would using `_TempDir()` be better?

christianbrodbeck commented:
Apart from this issue (which I'm not sure about) it looks goo to me

Eric89GXL commented:
`assert` lines are skipped when doing `python -O`. Also `assert_*` functions give more control in some cases, for example `assert_equal(a, b)` will give useful information about what a and b actually were compared to doing `assert a == b`.

(N.B. for projects using `py.test` it does some magic so `assert` statements are potentially preferred, but we use `nose` so it's not)

Eric89GXL commented:
Using `with NamedTemporaryFile(...)` will have problems on Windows if you then try to read the file, so you end up needing to do an annoying `with NamedTemporaryFile(..., delete=False)` and then manually delete it yourself later.

Instead you can just use `tempdir = _TempDir()` inside the function and it will auto-delete when `tempdir` gets GC'ed (i.e., at the end of the function)

kdoelling1919 commented:
Awesome. I just added the change to `_TempDir()`. 

Thanks a lot for your help with these reviews everyone! I just learned a ton about how to do this properly even on something that seemed relatively simple at the start. I really appreciate it.

Eric89GXL commented:
> I just learned a ton about how to do this properly even on something that seemed relatively simple at the start.

There are a lot of little tricks to keep things efficient and general, thanks for sticking with it

Eric89GXL commented on mne/io/kit/kit.py:
```diff
@@ -804,6 +804,11 @@ def read_raw_kit(input_fname, mrk=None, elp=None, hsp=None, stim='>',
     raw : Instance of RawKIT
         A Raw object containing KIT data.
 
+    Notes
```
Technically, `Notes` ]should come after] (https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt) `See Also`, but we don't need to hold up the PR for that

Eric89GXL commented on mne/io/kit/tests/test_kit.py:
```diff
@@ -36,6 +38,8 @@
 elp_path = op.join(data_dir, 'test.elp')
 hsp_path = op.join(data_dir, 'test.hsp')
+# this is a fake headshape made up of a unit sphere to test decimation
+sphere_hsp_txt_path = op.join(data_dir, 'test_sphere_hsp.txt')
```
This is no longer used so should be deleted. But we can fix this in a follow-up PR

Eric89GXL commented:
Thanks @kdoelling1919

Eric89GXL commented:
FYI I forgot to ask you to update `whats_new.rst` so I did it:

https://github.com/mne-tools/mne-python/commit/941694dc79196ed6e6e249979207c3b35e6a34e4

But it looks like your website (which I took from your GitHub account) is down
